packages/api/app/launchpadSaml.js

```diff
@@ -0,0 +1,127 @@
+  if (certificate) return flatten(certificate);
+  return [];
+};
+
+/**
+ * Store the SAML response's token in the AccessResponse table and return a JWT
+ * from the derived values.
+ *
+ * @param {Object} samlResponse - post_assert response from saml IDP provider
+ *
+ * @returns {Promise<Object>} - a valid JWT token that can be used for authentication.
+ */
+const buildLaunchpadJwtToken = async (samlResponse) => {
+  const {
+    user: { name_id: username, session_index: accessToken }
+  } = samlResponse;
+
+  const expirationTime = Date.now() + 60 * 60 * 1000;
+  const accessTokenModel = new AccessToken();
+  await accessTokenModel.create({ accessToken, expirationTime, username });
+  return createJwtToken({ accessToken, expirationTime, username });
+};
+
+/**
+* convenience function to set up SAML Identity and Service Providers
+*/
+const prepareSamlProviders = async () => {
+  const LaunchpadX509Certificate = await launchpadPublicCertificate(
+    process.env.LAUNCHPAD_METADATA_PATH
+  );
+
+  const spOptions = {
+    entity_id: process.env.ENTITY_ID,
+    assert_endpoint: process.env.ASSERT_ENDPOINT,
+    force_authn: false,
+    nameid_format: 'urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified',
+    sign_get_request: false,
+    allow_unencrypted_assertion: true
+  };
+
+  const idpOptions = {
+    sso_login_url: process.env.IDP_LOGIN,
+    sso_logout_url: null,
+    certificates: LaunchpadX509Certificate
+  };
+
+  const idp = new saml2.IdentityProvider(idpOptions);
+  const sp = new saml2.ServiceProvider(spOptions);
+
+  return { idp, sp };
+};
+
+// Starting point for SAML SSO login
+const login = async (req, res) => {
+  // saml2-js stuff
+  const { idp, sp } = await prepareSamlProviders();
+  const relayState = req.query.RelayState;
+  sp.create_login_request_url(
+    idp,
+    { relay_state: relayState },
+    (err, loginUrl) => {
+      if (err != null) return res.boom.badRequest('Could not create login request url.', err);
+      return res.redirect(loginUrl);
+    }
+  );
+};
+
+// SAML AssertionConsumerService (ACS) endpoint.
+const auth = async (req, res) => {
+  const { idp, sp } = await prepareSamlProviders();
+  sp.post_assert(idp, { request_body: req.body }, (err, samlResponse) => {
+    if (err != null) {
+      return res.boom.badRequest('SAML post assert error', err);
+    }
+    return buildLaunchpadJwtToken(samlResponse)
+      .then((LaunchpadJwtToken) => {
+        const Location = `${req.body.RelayState}/?token=${LaunchpadJwtToken}`;
+        return res.redirect(Location);
+      })
+      .catch((error) => res.boom.badRequest('Could not build JWT from SAML response', error));
+  });
+};
+
+const notImplemented = async (req, res) => res.boom.notImplemented(
+  `endpoint: "${req.path}" not implemented. Login with launchpad.`
+);
+
+const tokenEndpoint = notImplemented;
+const refreshEndpoint = notImplemented;
+const deleteTokenEndpoint = notImplemented;
```

Comment:
Why is this disabled? The delete token endpoint was added to clean up tokens when a user logs out/is logged out of the dashboard, so that we can be sure no one can still use those tokens. It seems like that would still be a valid concern in this implementation.

Reply:
Honestly because I haddn't looked at it, and assumed it was related to the token stuff above.  I could probably un-disable that in this pr.